### ISSUE TITLE
[BUG] User changes on reboot

### ISSUE THREAD

**Issue by elesbb** [ElderDrivers/EdXposed#654]
**什么情况/What happened?**
After a random reboot, (the reboot does not always trigger this), the device will act as if it has not been setup before (shows the android is starting prompt followed by the setup wizard). 

**Xposed模块列表/Xposed Module List**
NONE - Eliminating the fact it could have been a module

**Magisk模块列表/Magisk Module List**
Riru - Core
Magisk EdXposed

**EdXposed及Riru版本/Versions of EdXposed and Riru**

EdXposed:  0.4.6.4 (4563) (Sandhook)

Riru: 21.3

**相关Logcat/Logcat**
Hard to get a logcat because it happens during a reboot. I however did get a logcat AFTER the user changes and all the apps force close.

In the logcat, you can see all the crashes from facebook (me trying to open it after it reboots) and they are all the same:
`11-23 14:33:40.950 14391 14391 E AndroidRuntime: java.lang.RuntimeException: Unable to instantiate application com.facebook.lite.ClientApplicationSplittedShell: java.lang.IllegalArgumentException: Optimized data directory /data/user/0/com.facebook.lite/dex is not owned by the current user. Shared storage cannot protect your application from code injection attacks.`

I do know each app is installed with its own unique userId. So something is changing this after the reboot..

[facebook.txt](https://github.com/ElderDrivers/EdXposed/files/5590689/facebook.txt)

**Comment by elesbb** [comment]
XDA Post: https://forum.xda-developers.com/showpost.php?p=83983677&postcount=447

**Comment by kotori2** [comment]
`AndroidRuntime: 	at X.02Q.A00(:67)`
Doesn't seems like a EdXposed class. EdXposed uses in-memory dex loader so it won't produce any dex file on disk.
Also please paste related logcat here.

**Comment by elesbb** [comment]
@kotori2 
Hey, thanks for helping out with this. I understand that the class which is crashing is not related to EdXposed. The problem is EdXposed (or something else?) is somehow changing the owner of all installed apps (including system apps), which is identified by the ` Optimized data directory /data/user/0/com.facebook.lite/dex is not owned by the current user. Shared storage cannot protect your application from code injection attacks.` log. That is why the class `X.02Q.A00` crashes, because it cannot access it's own storage directory.

Related Logcat:
`11-23 14:33:40.830 14391 14391 W ContextImpl: Failed to ensure /data/user_de/0/com.facebook.lite/code_cache: mkdir failed: EACCES (Permission denied)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: FATAL EXCEPTION: main
11-23 14:33:40.950 14391 14391 E AndroidRuntime: Process: com.facebook.lite, PID: 14391
11-23 14:33:40.950 14391 14391 E AndroidRuntime: java.lang.RuntimeException: Unable to instantiate application com.facebook.lite.ClientApplicationSplittedShell: java.lang.IllegalArgumentException: Optimized data directory /data/user/0/com.facebook.lite/dex is not owned by the current user. Shared storage cannot protect your application from code injection attacks.
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.LoadedApk.makeApplication(LoadedApk.java:1296)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.ActivityThread.handleMakeApplication(ActivityThread.java:7152)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7134)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at com.swift.sandhook.SandHook.callOriginMethod(SandHook.java:185)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at com.swift.sandhook.xposedcompat.hookstub.HookStubManager.hookBridge(HookStubManager.java:376)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at SandHookerNew_5sm3dthec8858ee0to4kj3c1ge.hook(Unknown Source:51)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.ActivityThread.access$1600(ActivityThread.java:274)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2102)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:107)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:237)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8167)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:496)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1100)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: Caused by: java.lang.IllegalArgumentException: Optimized data directory /data/user/0/com.facebook.lite/dex is not owned by the current user. Shared storage cannot protect your application from code injection attacks.
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at dalvik.system.DexFile.<init>(DexFile.java:141)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at dalvik.system.DexFile.loadDex(DexFile.java:210)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at dalvik.system.DexFile.loadDex(DexFile.java:177)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at X.02Q.A00(:67)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at X.01U.A04(:61)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at com.facebook.lite.ClientApplicationSplittedShell.attachBaseContext(:4738)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.Application.attach(Application.java:370)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.Instrumentation.newApplication(Instrumentation.java:1157)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	at android.app.LoadedApk.makeApplication(LoadedApk.java:1288)
11-23 14:33:40.950 14391 14391 E AndroidRuntime: 	... 14 more`

**Comment by elesbb** [comment]
Also, `11-23 14:33:40.830 14391 14391 W ContextImpl: Failed to ensure /data/user_de/0/com.facebook.lite/code_cache: mkdir failed: EACCES (Permission denied)` shows it cannot access its own /data/data directory, because the owner has changed as the permissions for that directory are 700 (only owner had access)

PS: Thanks for editing the logcat, couldn't figure out how to do that.

**Comment by kotori2** [comment]
> The problem is EdXposed (or something else?) is somehow changing the owner of all installed apps (including system apps)

It definitely won't do that. Or you will see tons of SELinux error.
Also from the logs it seems Facebook is trying to load a hot patch and the permission is wrong. `code_cache` should be 771 in my case. Kinda sus some module/plugin lead to this error.

**Comment by elesbb** [comment]
> > The problem is EdXposed (or something else?) is somehow changing the owner of all installed apps (including system apps)
> 
> It definitely won't do that. Or you will see tons of SELinux error.
> Also from the logs it seems Facebook is trying to load a hot patch and the permission is wrong. `code_cache` should be 771 in my case. Kinda sus some module/plugin lead to this error.

Correct, `code_cache` is 771. However, `/data/user_de/0/com.facebook.lite` is 700. `/data/data/com.facebook.lite` is also 700. `ls -l` shows the owner of `/data/data/com.facebook.lite` is u0_a234. The calling owner from the facebook app is no longer u0_a234, this leads to the crash.

The above also happens to all the apps on the device, including system apps. When the system app cannot read it's respective /data/data or /data/user_de/0 directories, then it wipes them and re-creates them, now there is no reference that the device has been setup, prompting the "android is starting" message. 

In terms of permissions, all permissions are correct, the owner is just different now and cannot access the contents. I can do `chmod -R 777 /data/*` and most apps will open without a crash because the permissions slot for "other" is now full access.

I completely agree that this is sus and weird. I have done extensive googling for anyone else with this problem and not a single result was found. I have no idea if the system is doing this, or what. I know a lot of security measures were implemented in android 10. I thought I disabled most if not all of them tho.

**Comment by elesbb** [comment]
@kotori2 

Here is another relevant logcat from Textra, where it cannot open the messaging.db to read the SMS/MMS stored in it.

```
11-23 14:34:20.354  9346  9346 W ContextImpl: Failed to ensure /data/user/0/com.textra/cache: mkdir failed: EACCES (Permission denied)
11-23 14:34:20.354  9346  9346 W ContextImpl: Failed to update user.inode_cache: stat failed: EACCES (Permission denied)
11-23 14:34:20.355  9346  9346 W ContextImpl: Failed to ensure /data/user_de/0/com.textra/code_cache: mkdir failed: EACCES (Permission denied)
11-23 14:34:20.355  9346  9346 W ContextImpl: Failed to update user.inode_code_cache: stat failed: EACCES (Permission denied)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: Failed to open database '/data/user/0/com.textra/databases/messaging.db'.
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: android.database.sqlite.SQLiteCantOpenDatabaseException: unknown error (code 1806 SQLITE_CANTOPEN_EACCES[1806]): Could not open database
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteConnection.nativeOpen(Native Method)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:300)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteConnection.open(SQLiteConnection.java:218)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteConnectionPool.openConnectionLocked(SQLiteConnectionPool.java:737)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:284)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteConnectionPool.open(SQLiteConnectionPool.java:251)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteDatabase.openInner(SQLiteDatabase.java:1394)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteDatabase.open(SQLiteDatabase.java:1339)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteDatabase.openDatabase(SQLiteDatabase.java:974)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteOpenHelper.getDatabaseLocked(SQLiteOpenHelper.java:448)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at android.database.sqlite.SQLiteOpenHelper.getWritableDatabase(SQLiteOpenHelper.java:391)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at com.mplus.lib.lg1.d(SourceFile:1)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at com.mplus.lib.mg1.<init>(SourceFile:6)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at com.mplus.lib.mg1.j0(SourceFile:1)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at com.mplus.lib.af1.run(Unknown Source:2)
11-23 14:34:20.432  9346 26624 E SQLiteDatabase: 	at java.lang.Thread.run(Thread.java:919)
```

**Comment by kotori2** [comment]
So will the app user changes every reboot? It sounds like some multi-user issue.

**Comment by elesbb** [comment]
> So will the app user changes every reboot? It sounds like some multi-user issue.

I wish. It changes at random. Here is what I am doing that causes the change:

1. Install my module
2. Test it, okay need to change something
3. Changes it in android studio
4. Installs app via the "run" button in studio (test package is disabled, thought this might have been doing it but nope)
5. EdXposed manager says "reboot to apply update to module"
6. I reboot
7. All is fine
8. Start at 2 and go to 6 a random number of times
9. Now I am prompted with "Android is Starting" and then the device setup activity as if a factory reset was performed.
10. All my apps are still there, and their data, but the logcats above are generated with a force close of the app

I also have duplicate entries in my EdXposed manager log:
```
11-23 15:27:10.808  1160  1160 I EdXposed-Bridge: Loading modules from /data/app/com.elesbb.keyboardcustomizer-He6x2mynTtL8LNE9pzfvaA==/base.apk
11-23 15:27:10.846  1160  1160 I EdXposed-Bridge:   Loading class com.elesbb.keyboardcustomizer.XposedInit
**11-23 15:27:25.264  1161  1161 I EdXposed-Bridge: Loading modules from /data/app/com.elesbb.keyboardcustomizer-He6x2mynTtL8LNE9pzfvaA==/base.apk
11-23 15:27:25.372  1161  1161 I EdXposed-Bridge:   Loading class com.elesbb.keyboardcustomizer.XposedInit**
```

I looked in `ps -A | grep 1161` and each log above is zygote64 (1160) and zygote (1161) which are normal?

I only ever used xposed in the past and it only had one running not two, not sure if this is normal for EdXposed

**Comment by elesbb** [comment]
I too thought it was a multi-user thing, but I don't have any other user on the device?

**Comment by kotori2** [comment]
> I looked in ps -A | grep 1161 and each log above is zygote64 (1160) and zygote (1161) which are normal?

Yes it was normal. But it was changed since #655. 
> I don't have any other user on the device?

That's so weird because obviously I can't reproduce it. So I can only guessing. 
Riru does hooked nativeForkAndSpecialize and initialize EdXposed at that time, which might mess up with user id. But I still think the chance is too low.

**Comment by elesbb** [comment]
> Yes it was normal. But it was changed since #655.
Does the change mean it should only be running one? 
> > I don't have any other user on the device?
> 
> That's so weird because obviously I can't reproduce it. So I can only guessing.
> Riru does hooked nativeForkAndSpecialize and initialize EdXposed at that time, which might mess up with user id. But I still think the chance is too low.

I completely agree. I have been using xposed for since ever and this is the first time I have ever had issues with it. 

I am on a Sprint Samsung Galaxy S10 5G, and I don't know if Samsung has something that is messing with me but everything I could think of in terms of multi-user is gone.

**Comment by elesbb** [comment]
Well, thankfully it is NOT Xposed related. Just got the same problem without any riru or anything edxposed related installed. I wonder if I am debloating too much. 

Closing this. 

PS: I was not rigorous enough in my testing. I would go a week without xposed and no issues (but the amount of app installing/rebooting was not as high as when I would install xposed) Now I tested by basically pretending xposed was installed and kept installing /rebooting my module I was testing. 

Sorry for all the trouble and thank you for chatting with me!
